src/main/scala/mesosphere/marathon/core/instance/Instance.scala

```diff
@@ -1,20 +1,46 @@
 package mesosphere.marathon.core.instance
 
+import java.util.Base64
+
+import mesosphere.marathon.Protos
 import mesosphere.marathon.core.instance.Instance.InstanceState
 import mesosphere.marathon.core.task.Task
-import mesosphere.marathon.state.{ PathId, Timestamp }
+import mesosphere.marathon.state.{ MarathonState, PathId, Timestamp }
 import org.apache._
+import org.apache.mesos.Protos.Attribute
+import play.api.libs.json.{ Format, JsResult, JsString, JsValue, Json }
 
-case class Instance(instanceId: Instance.Id, agentInfo: Instance.AgentInfo, state: InstanceState, tasks: Seq[Task]) {
+// TODO: remove MarathonState stuff once legacy persistence is gone
+case class Instance(instanceId: Instance.Id, agentInfo: Instance.AgentInfo, state: InstanceState, tasks: Seq[Task])
+    extends MarathonState[Protos.Json, Instance] {
 
   def runSpecVersion: Timestamp = state.version
   def runSpecId: PathId = instanceId.runSpecId
 
   def isLaunched: Boolean = tasks.forall(task => task.launched.isDefined)
+
+  override def mergeFromProto(message: Protos.Json): Instance = {
+    Json.parse(message.getJson).as[Instance]
+  }
+  override def mergeFromProto(bytes: Array[Byte]): Instance = {
+    mergeFromProto(Protos.Json.parseFrom(bytes))
+  }
+  override def toProto: Protos.Json = {
+    Protos.Json.newBuilder().setJson(Json.stringify(Json.toJson(this))).build()
+  }
+  override def version: Timestamp = Timestamp.zero
 }
 
 object Instance {
 
+  import mesosphere.marathon.api.v2.json.Formats
```

Comment:
We should actually just delete Formats. The correct TODO would be to move Timestamp Json into the companion.